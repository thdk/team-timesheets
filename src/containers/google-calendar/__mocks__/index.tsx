import React from "react";

export const GoogleCalendarEvents = () => <>GoogleCalendarEvents</>;