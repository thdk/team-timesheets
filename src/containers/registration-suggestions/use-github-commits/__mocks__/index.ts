export const useGithubCommits = jest.fn().mockReturnValue({
    data: [],
    hasData: false,
});
