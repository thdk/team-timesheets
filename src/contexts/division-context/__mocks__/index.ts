export const useDivisionStore = jest.fn().mockReturnValue({
    userDivisions: [],
});
