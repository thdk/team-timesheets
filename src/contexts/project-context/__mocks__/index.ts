export const useProjectStore = jest.fn().mockReturnValue({
    projects: [],
    collection: {
        isFetched: true,
        docs: [],
    },
    activeProjects: [],
});
