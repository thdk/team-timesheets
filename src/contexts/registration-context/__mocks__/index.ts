export * from "./use-registration-store";
