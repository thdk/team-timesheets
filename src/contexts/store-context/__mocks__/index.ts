export const useStore = jest.fn();
