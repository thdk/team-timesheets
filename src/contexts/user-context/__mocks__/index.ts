export const useUserStoreMock = {
    authenticatedUser: {
        email: "foobar@email.com",
        name: "Foobar",
        roles: {
            admin: true,
        },
    },
    divisionUser: {
        email: "foobar@email.com",
        name: "Foobar",
        roles: {
            admin: true,
        },
    },
    usersCollection: {
        isFetched: true,
    },
    divisionUsersCollection: {
        isFetched: true,
    },
    updateDivisionUser: jest.fn(),
} as ReturnType<typeof useUserStore>;

export const useUserStore = jest.fn().mockReturnValue(useUserStoreMock);
