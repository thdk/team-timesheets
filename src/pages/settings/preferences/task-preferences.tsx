import React, { memo, useCallback } from "react";
import { TasksChips } from "../../../containers/tasks/chips";
import { useUserStore } from "../../../contexts/user-context";
import { useTasks } from "../../../contexts/task-context";

export const TaskPreferences = memo(() => {
    const userStore = useUserStore();
    const { tasks } = useTasks();
    const {
        tasks: userTasksIds = new Map<string, true>(),
        defaultTask = undefined,
    } = userStore.divisionUser || {};

    const handleTaskSelect = useCallback((id: string, selected: boolean) => {

        if (selected === !!userTasksIds.get(id)) return;

        if (selected) userTasksIds.set(id, true);
        else userTasksIds.delete(id);

        userStore.updateDivisionUser({
            tasks: userTasksIds,
        });
    }, [userTasksIds]);

    const defaultTaskChanged = useCallback((defaultTask: string) => {
        userStore.updateDivisionUser({
            defaultTask
        });
    }, [userStore]);

    return (
        <>
            <h3 className="mdc-typography--subtitle1">
                Pick your tasks
            </h3>
            <p>Only selected tasks will be available for you when adding a new regisration.</p>
            <TasksChips
                tasks={tasks}
                onTaskInteraction={(id, selected) => handleTaskSelect(id, !selected)}
                selectedTaskIds={Array.from(userTasksIds.keys())}
                filter
            />

            <h3 className="mdc-typography--subtitle1">Pick your default task</h3>
            <p>This task will be selected by default when you create a new registration.</p>
            <TasksChips
                tasks={tasks.filter((t => userTasksIds.get(t.id)))}
                onTaskInteraction={defaultTaskChanged}
                selectedTaskIds={defaultTask ? [defaultTask] : undefined}
                choice
            />
        </>
    );
});
