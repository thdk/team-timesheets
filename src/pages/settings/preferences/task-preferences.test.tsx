import React from "react";
import { render, fireEvent } from "@testing-library/react";

import { ITask } from "../../../../common";
import { useUserStore } from "../../../contexts/user-context";
import { useUserStoreMock } from "../../../contexts/user-context/__mocks__";
import { TaskPreferences } from "./task-preferences";
import { useTasks } from "../../../contexts/task-context";

jest.mock('../../../contexts/user-context');
jest.mock('../../../contexts/task-context');

const resetMocks = () => {
    (useTasks as jest.Mock<ReturnType<typeof useTasks>>).mockReturnValue({
        tasks: [] as ITask[],
    } as ReturnType<typeof useTasks>);

    (useUserStore as jest.Mock<ReturnType<typeof useUserStore>>).mockReturnValue(
        useUserStoreMock
    );
}

beforeEach(resetMocks);
afterAll(resetMocks);

describe("TaskPreferences", () => {
    it("should not render without tasks", () => {
        const { asFragment } = render(
            <TaskPreferences />
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("should not render without authenticated user", () => {
        (useUserStore as jest.Mock<ReturnType<typeof useUserStore>>).mockReturnValue(
            {
                ...useUserStoreMock,
                divisionUser: undefined,
            }
        );
        const { asFragment } = render(
            <TaskPreferences />
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("should render tasks", () => {
        (useTasks as jest.Mock<ReturnType<typeof useTasks>>).mockReturnValue({
            tasks: [
                {
                    id: "task-1",
                    name: "Task 1",
                },
                {
                    id: "task-2",
                    name: "Task 2",
                },
            ],
        } as ReturnType<typeof useTasks>);

        const { asFragment } = render(
            <TaskPreferences />
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("should show default task", () => {
        (useUserStore as jest.Mock<ReturnType<typeof useUserStore>>).mockReturnValue(
            {
                ...useUserStoreMock,
                divisionUser: {
                    ...useUserStoreMock.divisionUser,
                    defaultTask: "task-1",
                    tasks: new Map([["task-1", true], ["task-2", true]]),
                }
            }
        );

        (useTasks as jest.Mock<ReturnType<typeof useTasks>>).mockReturnValue({
            tasks: [
                {
                    id: "task-1",
                    name: "Task 1",
                },
                {
                    id: "task-2",
                    name: "Task 2",
                },
            ],
        } as ReturnType<typeof useTasks>);

        const { asFragment } = render(
            <TaskPreferences />
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("should call updateDivisionUser when default task is changed", () => {
        const updateDivisionUser = jest.fn();

        const originalMock = jest.requireMock('../../../contexts/user-context').useUserStore();
        (useUserStore as jest.Mock<ReturnType<typeof useUserStore>>).mockReturnValue(
            {
                ...originalMock,
                updateDivisionUser,
                divisionUser: {
                    ...originalMock.divisionUser,
                    tasks: new Map([["task-1", true], ["task-2", true]]),
                },
            }
        );

        (useTasks as jest.Mock<ReturnType<typeof useTasks>>).mockReturnValue({
            tasks: [
                {
                    id: "task-1",
                    name: "Task 1",
                },
                {
                    id: "task-2",
                    name: "Task 2",
                },
            ],
        } as ReturnType<typeof useTasks>);

        const { container } = render(
            <TaskPreferences />
        );

        const chipEl = Array.from(
            container.querySelectorAll(".mdc-chip-set--choice .mdc-chip__text")
        ).find(el => el.innerHTML === "Task 1");

        fireEvent.click(chipEl!);

        expect(updateDivisionUser).toHaveBeenCalledWith({
            defaultTask: "task-1",
        });
    });

    it("should call updateDivisionUser when task is added to filter", () => {
        const updateDivisionUser = jest.fn();
        const originalMock = jest.requireMock('../../../contexts/user-context').useUserStore();
        (useUserStore as jest.Mock<ReturnType<typeof useUserStore>>).mockReturnValue(
            {
                ...originalMock,
                updateDivisionUser,
            }
        );

        (useTasks as jest.Mock<ReturnType<typeof useTasks>>).mockReturnValue({
            tasks: [
                {
                    id: "task-1",
                    name: "Task 1",
                },
                {
                    id: "task-2",
                    name: "Task 2",
                },
            ],
        } as ReturnType<typeof useTasks>);

        const { getByText } = render(
            <TaskPreferences />
        );

        fireEvent.click(getByText("Task 1"));

        expect(updateDivisionUser).toHaveBeenCalledWith({
            tasks: new Map<string, true>([["task-1", true]]),
        });
        expect(updateDivisionUser).toHaveBeenCalledTimes(1);

        // Call it twice to see if it only run once
        fireEvent.click(getByText("Task 1"));
        expect(updateDivisionUser).toHaveBeenCalledTimes(1);
    });

    it("should call updateDivisionUser when task is removed from filter", () => {
        const updateDivisionUser = jest.fn();
        const originalMock = jest.requireMock('../../../contexts/user-context').useUserStore();
        (useUserStore as jest.Mock<ReturnType<typeof useUserStore>>).mockReturnValue(
            {
                ...originalMock,
                updateDivisionUser,
                divisionUser: {
                    ...originalMock.divisionUser,
                    tasks: new Map([["task-1", true]])
                },
            }
        );

        (useTasks as jest.Mock<ReturnType<typeof useTasks>>).mockReturnValue({
            tasks: [
                {
                    id: "task-1",
                    name: "Task 1",
                },
                {
                    id: "task-2",
                    name: "Task 2",
                },
            ],
        } as ReturnType<typeof useTasks>);

        const { getByText } = render(
            <TaskPreferences />
        );

        fireEvent.click(getByText("Task 1"));

        expect(updateDivisionUser).toHaveBeenCalledWith({
            tasks: new Map(),
        });
    });

    // it("should call updateDivision user when client is changed", async () => {
    //     const updateDivisionUser = jest.fn();
    //     (useUserStore as jest.Mock<ReturnType<typeof useUserStore>>).mockReturnValue(
    //         {
    //             ...useUserStoreMock,
    //             updateDivisionUser
    //         },
    //     );

    //     (useClients as jest.Mock<ReturnType<typeof useClients>>).mockReturnValue({
    //         clients: [
    //             {
    //                 id: "client-1",
    //                 name: "Client 1",
    //             },
    //             {
    //                 id: "client-2",
    //                 name: "Client 2",
    //             },
    //         ],
    //     } as ReturnType<typeof useClients>);

    //     const { container, getByText } = render(
    //         <ClientPreferences />
    //     );

    //     const selectEl = container.querySelector("select");
    //     expect(selectEl).toBeDefined();

    //     userEvent.selectOptions(selectEl!, getByText("Client 2"));

    //     await waitFor(() => expect(updateDivisionUser).toBeCalled());
    // });
});
