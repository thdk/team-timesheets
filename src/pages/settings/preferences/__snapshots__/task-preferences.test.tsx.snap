// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskPreferences should not render without authenticated user 1`] = `
<DocumentFragment>
  <h3
    class="mdc-typography--subtitle1"
  >
    Pick your tasks
  </h3>
  <p>
    Only selected tasks will be available for you when adding a new regisration.
  </p>
  <div
    class="mdc-chip-set mdc-chip-set--filter"
  />
</DocumentFragment>
`;

exports[`TaskPreferences should not render without tasks 1`] = `
<DocumentFragment>
  <h3
    class="mdc-typography--subtitle1"
  >
    Pick your tasks
  </h3>
  <p>
    Only selected tasks will be available for you when adding a new regisration.
  </p>
  <div
    class="mdc-chip-set mdc-chip-set--filter"
  />
</DocumentFragment>
`;

exports[`TaskPreferences should render tasks 1`] = `
<DocumentFragment>
  <h3
    class="mdc-typography--subtitle1"
  >
    Pick your tasks
  </h3>
  <p>
    Only selected tasks will be available for you when adding a new regisration.
  </p>
  <div
    class="mdc-chip-set mdc-chip-set--filter"
  >
    <div
      class="mdc-ripple-surface mdc-chip"
      id="task-1"
      role="row"
    >
      <div
        class="mdc-chip__ripple"
      />
      <div
        class="mdc-chip__checkmark"
      >
        <svg
          class="mdc-chip__checkmark-svg"
          viewBox="-2 -3 30 30"
        >
          <path
            class="mdc-chip__checkmark-path"
            d="M1.73,12.91 8.1,19.28 22.79,4.59"
            fill="none"
            stroke="black"
          />
        </svg>
      </div>
      <span
        role="gridcell"
      >
        <span
          class="mdc-chip__text__primary-action"
          role="button"
          tabindex="0"
        >
          <span
            class="mdc-chip__text"
          >
            Task 1
          </span>
        </span>
      </span>
    </div>
    <div
      class="mdc-ripple-surface mdc-chip"
      id="task-2"
      role="row"
    >
      <div
        class="mdc-chip__ripple"
      />
      <div
        class="mdc-chip__checkmark"
      >
        <svg
          class="mdc-chip__checkmark-svg"
          viewBox="-2 -3 30 30"
        >
          <path
            class="mdc-chip__checkmark-path"
            d="M1.73,12.91 8.1,19.28 22.79,4.59"
            fill="none"
            stroke="black"
          />
        </svg>
      </div>
      <span
        role="gridcell"
      >
        <span
          class="mdc-chip__text__primary-action"
          role="button"
          tabindex="0"
        >
          <span
            class="mdc-chip__text"
          >
            Task 2
          </span>
        </span>
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TaskPreferences should show default task 1`] = `
<DocumentFragment>
  <h3
    class="mdc-typography--subtitle1"
  >
    Pick your tasks
  </h3>
  <p>
    Only selected tasks will be available for you when adding a new regisration.
  </p>
  <div
    class="mdc-chip-set mdc-chip-set--filter"
  >
    <div
      class="mdc-ripple-surface mdc-chip mdc-chip--selected"
      id="task-1"
      role="row"
    >
      <div
        class="mdc-chip__ripple"
      />
      <div
        class="mdc-chip__checkmark"
      >
        <svg
          class="mdc-chip__checkmark-svg"
          viewBox="-2 -3 30 30"
        >
          <path
            class="mdc-chip__checkmark-path"
            d="M1.73,12.91 8.1,19.28 22.79,4.59"
            fill="none"
            stroke="black"
          />
        </svg>
      </div>
      <span
        role="gridcell"
      >
        <span
          class="mdc-chip__text__primary-action"
          role="button"
          tabindex="0"
        >
          <span
            class="mdc-chip__text"
          >
            Task 1
          </span>
        </span>
      </span>
    </div>
    <div
      class="mdc-ripple-surface mdc-chip mdc-chip--selected"
      id="task-2"
      role="row"
    >
      <div
        class="mdc-chip__ripple"
      />
      <div
        class="mdc-chip__checkmark"
      >
        <svg
          class="mdc-chip__checkmark-svg"
          viewBox="-2 -3 30 30"
        >
          <path
            class="mdc-chip__checkmark-path"
            d="M1.73,12.91 8.1,19.28 22.79,4.59"
            fill="none"
            stroke="black"
          />
        </svg>
      </div>
      <span
        role="gridcell"
      >
        <span
          class="mdc-chip__text__primary-action"
          role="button"
          tabindex="0"
        >
          <span
            class="mdc-chip__text"
          >
            Task 2
          </span>
        </span>
      </span>
    </div>
  </div>
  <h3
    class="mdc-typography--subtitle1"
  >
    Pick your default task
  </h3>
  <p>
    This task will be selected by default when you create a new registration.
  </p>
  <div
    class="mdc-chip-set mdc-chip-set--choice"
  >
    <div
      class="mdc-ripple-surface mdc-chip mdc-chip--selected"
      id="task-1"
      role="row"
    >
      <div
        class="mdc-chip__ripple"
      />
      <span
        role="gridcell"
      >
        <span
          class="mdc-chip__text__primary-action"
          role="button"
          tabindex="0"
        >
          <span
            class="mdc-chip__text"
          >
            Task 1
          </span>
        </span>
      </span>
    </div>
    <div
      class="mdc-ripple-surface mdc-chip"
      id="task-2"
      role="row"
    >
      <div
        class="mdc-chip__ripple"
      />
      <span
        role="gridcell"
      >
        <span
          class="mdc-chip__text__primary-action"
          role="button"
          tabindex="0"
        >
          <span
            class="mdc-chip__text"
          >
            Task 2
          </span>
        </span>
      </span>
    </div>
  </div>
</DocumentFragment>
`;
